### DISCUSSION TITLE
Spalten mit Hyperlinks fehlen in CSV-Export

### DISCUSSION THREAD

**Discussion by Andreas-Lampert** [e-netze/webgis-community#138] [General]
Hallo Jürgen!

Aufgefallen in Version 5.23.5101.

Beim Export von Abfrageergebnissen als CSV oder MS Excel (CSV) werden Spalten, die einen Hyperlink beinhalten, nicht exportiert. 

![csv_export](https://github.com/gis-eni/webgis-forum/assets/122986367/fc6c40d1-ee62-4190-92bb-e659e3347735)

CSV
GebietsNr.;GebietsName;Code (EU)
5;Wachau - Jauerling;AT1205000
1;Waldviertel;AT1201000

Excel CSV
GebietsNr.;GebietsName;Code (EU)
5;="Wachau - Jauerling";="AT1205000"
1;="Waldviertel";="AT1201000"

Du kannst das Beispiel bei uns im NÖ Atlas testen --> https://atlas.noe.gv.at/atlas/portal/noe-atlas/map/Naturraum/Naturschutz
Bei der Abfrage nach "Naturschutz - Natura 2000 Vogelschutzgebiete" sind Spalten mit Hyperlinks vorhanden, die nicht exportiert werden.

Der Export hat schon funktioniert. Muss in der Konfig etwas geändert werden, damit der Export der Link-Spalten wieder funktioniert?

Danke!

Lg, Andi

**Comment by gis-eni** [comment]
Hallo Andi,

ich kann das grundsätzlich nachvollziehen. Allerdings "funktioniert" das wie es ausschaut schon länger nicht nicht mehr. Im Code ist das bei mir absichtlich so markiert:

``` csharp
var columns = this.FirstOrDefault()
    .Attributes?
    .Where(a => !a.Name.StartsWith("_") && !a.Value.StartsWith("<a "))   // keine Links
    .Select(a => a.Name);
```

Also: Spalten mit Links werden absichtlich für den Export ausgenommen. Laut SourceCode Verwaltung geht das bis 2021 (und länger zurück)
Die Frage ist, warum das so gemacht wurde? Es war glaub ich eher ein Security Feature, weil man Links nicht irgendwem weitergeben möchte. Oft werden diese Tabellen exportiert und dann weiter verschickt und dann kommt es zu folgenden Problemen:

* Links können auf Intranet Ziele zeigen oder enthalten Tokens, die man nicht nach außen weitergeben möchte/darf
* Links enthalten Keys (Felder) die eigentlich nicht sichtbar sind und Informationen enthalten, die man nicht weitergeben sollte/darf
* Links sind oft nur eine kurze Zeit gültig. Die Keys/Ids in den Parametern können sich täglich ändern. Die Links werden dadurch unbrauchbar oder erzeugen irgendwann falsche Ergebnisse

Darum möchte ich diese Feature nicht wieder rückgängig machen.

Einen Workaround kann ich noch anbieten, wenn du das für ein Thema unbedingt brauchst:

Du kannst bei jeder Abfrage im CMS sogenannte ``Exportierbare Formate`` anlegen. Damit kannst du aus deinen Daten beliebige Textformate exportieren. Für ein CSV mit Links könnte das zB so aussehen:

![image](https://github.com/gis-eni/webgis-forum/assets/62870568/fea5542f-6cb3-45e0-92d2-bff9f379673a)

In den eckigen Klammen, kannst du wieder Platzhalter für deine Felder angeben. Weil es CSV sein sollte kannst du als Platzhalter zwischen den Splaten `;` verwenden. Am Ende einer Zeile musst du einen Zeilenumbruch angeben `\n` (Das wird absichtlich nicht automatisch gemacht! 
Das Feld [GNR] (=Grundstücksnummer) schreib ich beispielsweise in folgender Form ins CSV: ``="10/1"``. Das macht Sinn, wenn du die Daten hauptsächlich ins Excel exportieren möchtest. Damit verhinderst du, dass Excel 10/1 als Datum (10. Jän) interpretiert. Es ist also damit so gut wie alles möglich.

Das Ergebnis ist ein Textfile mit der Endung CSV. Eine Zeile pro Feature. Allerdings fehlen die Überschriften, schon klar...

**Comment by Andreas-Lampert** [comment]
Hallo Jürgen!

Danke für die Info und den Workaround! Werde ich so konfigurieren.

Wäre es prinzipiell technisch möglich in einer zukünftigen WebGIS-Version (WebGIS 6) eine Konfig-Möglichkeit im CMS zu bekommen, welche Attribute exportiert werden können - z.B. Exportierbar ja/nein. Dann könnte der Admin einfach entscheiden, ob das Feld exportiert werden soll oder nicht. 
Unklar ob Bedarf an so einer Funktion besteht. Vielleicht können wir das beim WebGIS Wednesday mal kurz ansprechen. 

Lg, Andi
